File: CS330/4_test_project_in_java/Staff.java
/**
 * Staff members in the company
 */
public class Staff implements Comparable<Staff>, Cloneable {
	public Staff ()
	{
		theSurname = "";
		theGivenName = "";
	}

	public Staff (String surname, String givenName)
	{
		theSurname = surname;
		theGivenName = givenName;
	}

	public String getSurname()             {return theSurname;}
	public void putSurname(String surname) {theSurname = surname;}

	public String getGivenName()               {return theGivenName;}
	public void putGivenName(String givenName) {theGivenName = givenName;}


	public boolean equals (Object o)
	{
		Staff r = (Staff)o;
		return theSurname.equals(r.theSurname) 
		&& theGivenName.equals(r.theGivenName);
	}
	
	public int hashCode()
	{
		return 3 * theSurname.hashCode() + theGivenName.hashCode();
	}

	public int compareTo (Staff r)
	{	
		return toString().compareTo(r.toString());
	}

	public String toString()
	{
		return theSurname + ", " + theGivenName;
	}
	
	public Object clone()
	{
		return new Staff(theSurname, theGivenName);
	}

	private String theSurname;
	private String theGivenName;
}

File: CS330/1_operator_overloading/staff.cpp
#include "staff.h"

using namespace std;

Staff::Staff ()
{}


Staff::Staff (std::string surname, std::string givenName)
  : theSurname (surname), theGivenName(givenName)
{}

std::ostream& operator<< (std::ostream& out, const Staff& s)
{
  out << s.getSurname();
  if (s.getGivenName().length() > 0)
    out << ", " << s.getGivenName();
  return out;
}

bool Staff::operator== (const Staff& r) const
{
  return theSurname == r.theSurname
    && theGivenName == r.theGivenName;
}

bool Staff::operator< (const Staff& r) const
{
  if (theSurname < r.theSurname)
    return true;
  if (theSurname > r.theSurname)
    return false;
  return (theGivenName < r.theGivenName);
}
